zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/JSONParser.scala

```diff
@@ -89,83 +111,112 @@ object JSONParser {
-      (JsPath \ "as").read[Seq[String]]
-  }.apply(LookupStatement.apply _)
-
-  implicit val unnestReads: Reads[Seq[UnnestStatement]] = new Reads[Seq[UnnestStatement]] {
-    override def reads(json: JsValue): JsResult[Seq[UnnestStatement]] = {
+  implicit val selectFormat: Format[SelectStatement] = (
+    (JsPath \ "order").format[Seq[String]] and
+      (JsPath \ "limit").format[Int] and
+      (JsPath \ "offset").format[Int] and
+      (JsPath \ "field").formatNullable[Seq[String]].inmap[Seq[String]](
+        o => o.getOrElse(Seq.empty[String]),
+        s => if (s.isEmpty) None else Some(s)
+      )
+    ) (SelectStatement.apply, unlift(SelectStatement.unapply))
+
+  implicit val lookupFormat: Format[LookupStatement] = (
+    (JsPath \ "sourceKey").format[Seq[String]] and
+      (JsPath \ "dataset").format[String] and
+      (JsPath \ "lookupKey").format[Seq[String]] and
+      (JsPath \ "select").format[Seq[String]] and
+      (JsPath \ "as").format[Seq[String]]
+    ) (LookupStatement.apply, unlift(LookupStatement.unapply))
+
+  implicit val unnestFormat: Format[UnnestStatement] = new Format[UnnestStatement] {
+    override def reads(json: JsValue): JsResult[UnnestStatement] = {
       JsSuccess(json.as[JsObject].value.map {
         case (key, jsValue: JsValue) =>
           UnnestStatement(key, jsValue.as[String])
-      }.toSeq)
+      }.head)
+    }
+
+    override def writes(unnestStatement: UnnestStatement): JsValue = {
+      JsObject(Seq(unnestStatement.fieldName -> JsString(unnestStatement.as)))
     }
   }
 
-  implicit val filterReads: Reads[FilterStatement] = {
-    (JsPath \ "field").read[String] and
-      (JsPath \ "apply").readNullable[TransformFunc] and
-      (JsPath \ "relation").read[Relation] and
-      (JsPath \ "values").read[Seq[Any]]
-  }.apply(FilterStatement.apply _)
+  implicit val filterFormat: Format[FilterStatement] = (
+    (JsPath \ "field").format[String] and
+      (JsPath \ "apply").formatNullable[TransformFunc] and
+      (JsPath \ "relation").format[Relation] and
+      (JsPath \ "values").format[Seq[Any]]
+    ) (FilterStatement.apply, unlift(FilterStatement.unapply))
 
   // TODO find better name for 'global'
-  implicit val queryReads: Reads[Query] = {
-    (JsPath \ "dataset").read[String] and
-      (JsPath \ "lookup").readNullable[Seq[LookupStatement]].map(_.getOrElse(Seq.empty)) and
-      (JsPath \ "filter").readNullable[Seq[FilterStatement]].map(_.getOrElse(Seq.empty)) and
-      (JsPath \ "unnest").readNullable[Seq[UnnestStatement]].map(_.getOrElse(Seq.empty)) and
-      (JsPath \ "group").readNullable[GroupStatement] and
-      (JsPath \ "select").readNullable[SelectStatement]  and
-      (JsPath \ "global").readNullable[GlobalAggregateStatement]
-  }.apply(Query.apply _)
+  implicit val queryFormat: Format[Query] = (
+    (JsPath \ "dataset").format[String] and
+      (JsPath \ "lookup").formatNullable[Seq[LookupStatement]].inmap[Seq[LookupStatement]](
```

Comment:
def inmap[B](f: A => B, g: B => A): M[B] = fu.inmap(ma, f, g)

It does what map does, but the opposite way too. I used it here, since we are not only reading, but also writing.
Got it from this answer: [http://stackoverflow.com/questions/29203069/format-nullable-seq-or-list-of-objects-with-play-json-and-salat]